File: apps/qizhi/mobile/MentorAI-Android/app/src/main/kotlin/com/mentorai/app/chat/ChatAPI.kt
package com.mentorai.app.chat

import com.mentorai.app.networking.APIClient
import com.mentorai.app.networking.AuthError
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.coroutines.flow.transformWhile
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

class ChatAPI(private val client: APIClient) {

    /**
     * POST /ai/chat — server replies with SSE. We build the request manually (POST + JSON
     * body + Accept: text/event-stream) and bridge OkHttp-SSE to a Flow of `ChatEvent`.
     *
     * Path matches the backend route (server `api/ai.py`, mounted under `/ai`) and the web
     * client (`postStream('/ai/chat', …)`). The request body shape already matched; the old
     * `/chat/send` path 404'd on this backend.
     */
    fun send(request: ChatSendRequest, token: String): Flow<ChatEvent> = flow {
        val payload = client.json.encodeToString(ChatSendRequest.serializer(), request)
        val httpRequest = client.buildRequest(
            method = "POST",
            path = "/ai/chat",
            body = client.jsonBody(payload),
            bearerToken = token,
            accept = "text/event-stream",
        )
        // Mirror iOS Chat/ChatAPI.swift: break on the first End/Error frame so a trailing error
        // followed by extra frames is never forwarded. transformWhile emits the terminal event,
        // then halts the upstream SSE collection.
        emitAll(
            SSEStream.messages(httpRequest, client.httpClient)
                .mapNotNull { sse -> SSEStream.parseChatEvent(sse, client.json) }
                .transformWhile { event ->
                    emit(event)
                    event !is ChatEvent.End && event !is ChatEvent.Error
                },
        )
    }
}

File: apps/qizhi/mobile/MentorAI-iOS/MentorAI/Chat/ChatAPI.swift
import Foundation
import os

struct ChatAPI {
    let baseURL: URL
    let session: URLSession
    let log = Logger(subsystem: "com.mentorai.app", category: "ChatAPI")

    init(baseURL: URL) {
        self.baseURL = baseURL
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 300
        configuration.timeoutIntervalForResource = 600
        self.session = URLSession(configuration: configuration)
    }

    init(baseURL: URL, session: URLSession) {
        self.baseURL = baseURL
        self.session = session
    }

    func send(_ request: ChatSendRequest, token: String) -> AsyncThrowingStream<ChatEvent, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    // Backend chat route is POST /ai/chat (server `api/ai.py`, mounted under
                    // the `/ai` prefix); the web client posts to the same path. The request
                    // body shape (query / session_id / file_paths / extra_params) already
                    // matches it — only the path was wrong, which 404'd on this backend.
                    let url = try makeURL(path: "ai/chat")
                    var urlRequest = URLRequest(url: url)
                    urlRequest.httpMethod = "POST"
                    urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    urlRequest.setValue("text/event-stream", forHTTPHeaderField: "Accept")
                    urlRequest.setValue("no-cache", forHTTPHeaderField: "Cache-Control")
                    urlRequest.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
                    let encoder = JSONEncoder()
                    let bodyData = try encoder.encode(request)
                    urlRequest.httpBody = bodyData
                    urlRequest.timeoutInterval = 300

                    let bodyPreview = String(data: bodyData, encoding: .utf8) ?? "<binary>"
                    log.debug("POST \(url.absoluteString, privacy: .public) body=\(bodyPreview, privacy: .public)")

                    for try await sse in SSEStream.messages(for: urlRequest, session: session) {
                        if Task.isCancelled { break }
                        guard let event = SSEStream.parseChatEvent(sse) else { continue }
                        continuation.yield(event)
                        if case .end = event { break }
                        if case .error = event { break }
                    }
                    continuation.finish()
                } catch is CancellationError {
                    continuation.finish(throwing: CancellationError())
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in
                task.cancel()
            }
        }
    }

    private func makeURL(path: String) throws -> URL {
        let trimmed = path.hasPrefix("/") ? String(path.dropFirst()) : path
        let base = baseURL.absoluteString.hasSuffix("/") ? baseURL.absoluteString : baseURL.absoluteString + "/"
        guard let url = URL(string: base + trimmed) else {
            throw AuthError.transport("Invalid URL for path \(path).")
        }
        return url
    }
}
